_posts/2018-09-05-Data-Continuity-Service-for-DynamoDB.md

```diff
@@ -0,0 +1,570 @@
+
+![configure lambda 2](/assets/posts/2018-09-05-Data-Continuity-Service-for-DynamoDB/configure-lambda-trigger-2.png)
+
+Or update your CloudFormation template with the content below below:
+```json
+{
+  "DynamoDbStreamToS3LambdaEventSourceMapping": {
+    "Properties": {
+      "EventSourceArn": "arn:aws:dynamodb:<YOUR_AWS_REGION>:<YOUR_AWS_ACCOUNT_ID>:table/users/stream/<STREAM_ID>",
+      "FunctionName": {
+        "Ref": "DynamoDbStreamToS3LambdaFunction"
+      },
+      "StartingPosition": "LATEST"
+    },
+    "Type": "AWS::Lambda::EventSourceMapping"
+  }
+}
+```
+
+Deployed? Done? Good. Everything on the Lambda side is ready apart from the function itself. Luckily the function itself is rather simple:
+
+```javascript
+'use strict';
+
+const aws = require('aws-sdk');
+
+const s3 = new aws.S3({ region: '<YOUR AWS_REGION>' });
+
+module.exports.backup = (event, context, callback) => {
+  const records = event.Records;
+
+  Promise.all(records.map((record) => {
+    const keysList = Object.keys(record.dynamodb.Keys).map((key) => {
+      const keyDefinition = record.dynamodb.Keys[key];
+      const type = Object.keys(keyDefinition)[0];
+      const value = keyDefinition[type];
+      return value;
+    });
+
+    const keysString = keysList[0];
+    let image = {}
+    if(typeof record.dynamodb.NewImage !== 'undefined') {
+      image = aws.DynamoDB.Converter.output({ M: record.dynamodb.NewImage });
+    }
+
+    return s3.putObject({
+      Bucket: process.env.BUCKET,
+      Key: `${process.env.TABLE_NAME}/${keysString}/image.json`,
+      Body: JSON.stringify(image),
+      StorageClass: "STANDARD_IA"
+    }).promise()
+      .then((response) => {
+        console.log(`${keysString} snapshot done`, response);
+      })
+      .catch((err) => {
+        console.error('Error', err);
+      });
+  }))
+    .then(v => callback(null, v), callback);
+};
+```
+
+Update the Lambda function with above code. Make sure you have set the Lambda handler to `index.backup`.
+
+In general the function will run for each DynamoDB __insert__ or __update__ action and read the event record. It will then write the "NewImage" data as a JSON file to S3 and store it as the *STANDARD_IA* storage class to save up some bucks as you most probably never have to use the data. Or as least that is what we are hoping for. :sweat_smile:
+
+It expects an environment variable named `TABLE_NAME` which holds the DynamoDB table name. In this case the table is called `users` so we are sending that with the function. It also expects another environment variable named `BUCKET`. Ahhh finally! We haven't really talked about buckets and S3 up to now so this would be a good moment to start talking about S3.
+
+S3
+-------
+This is where is really becomes interesting as we have to make sure the data is written to S3 and S3 is replicated into another region __and__ to another account.
+
+To do this you will have to configure two roles and two buckets. Both in different accounts each.
```

Comment:
To do this you will have to configure two roles and two buckets, both in different accounts each.

Reply:
"both" and "each" are redundant. I'd change it into:

> To do this you will have to configure two roles and two buckets. We will put one role and bucket in the source account, and the other role and bucket in the account containing the backups.

Reply:
Thanks